### ISSUE TITLE
Increase in memory consumption when fetching all documents from a database

### ISSUE THREAD

**Issue by dtto86** [IBM/cloudant-node-sdk#596]
**Describe the bug**
We observed increase in memory consumption when fetch 50 documents from a database using postAllDocs() function with parameters as 
{ db: 'abcd', includeDocs: true, limit: 50, skip: 0 }. We tried following scenarios

1. Executed script once and kept it running for 30 mins.

2. Executed script thrice with 5 mins of interval

Tried to fetch documents using postAllDocsAsStream() function with parameters { db: 'abcd', includeDocs: true, limit: 50, skip: 0 }

Can we consider this as a memory leak?

**To Reproduce**
Use postAllDocs() / postAllDocsAsStream() functions to reproduce the issue

**Expected behavior**
Memory consumption graph should come down after data retrieval.

**Screenshots**
1. Executed script once and kept it running for 30 mins.
![image](https://user-images.githubusercontent.com/5002000/144859563-fa4c4b26-3c46-4073-bfd0-45d3255d7b3a.png)

2. Executed script thrice with 5 mins of interval
![image](https://user-images.githubusercontent.com/5002000/144859639-6cfa3e88-23a4-4573-b5d3-ff04a2f157e5.png)

3. postAllDocsAsStream() function with parameters { db: 'abcd', includeDocs: true, limit: 50, skip: 0 }
![image](https://user-images.githubusercontent.com/5002000/144859722-bd7dcedf-0b0a-4c1b-8165-858a5f39fdd6.png)

**Must gather (please complete the following information):**
 - SDK Version - ^2.17.0
 - Node.js Version - 14.16.0

**Comment by dtto86** [comment]
In package.json, we have
`
"ibm-cloud-sdk-core": "^2.17.0",
"@ibm-cloud/cloudant": "0.0.22",
`

We used `BasicAuthenticator` to setup a connection and the size of 50 documents comes around 11346 bytes.

As you have noticed in all three attached screenshots, RSS starts around level 40 and it crosses level 60 within 10 mins and it stays there never jumps down to the initial level. This kind of behaviour makes us think that there is a memory leak.

**Comment by eiri** [comment]
HI @dtto86 

No, I wouldn't consider this sdk memory leak.

RSS (resident set size) is the portion of the memory allocated for the node process by system and it's not unusual for it to grow and stay arbitrary high on busy node process. Once system gets memory pressure it'll signal the node to free the unused "reserved" memory and node should run gc to do so.

Now If node can't release memory when asked that would be a memory leak, so I ran the following test based on your snippet:

`package.json`
```json
{
  "name": "express-cloudant",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "start": "node --trace-gc --expose-gc index.js",
    "loadtest": "k6 run --vus 5 --duration 5m load-test.js"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@ibm-cloud/cloudant": "0.0.22",
    "express": "^4.17.1"
  }
}
```

`index.js`
```js
const express = require('express');

const { memoryUsage } = require('process');

const { BasicAuthenticator } = require('ibm-cloud-sdk-core');
const { CloudantV1 } = require('@ibm-cloud/cloudant');

const app = express();
const username = "xxx";
const password = "xxx";
const serviceUrl = "http://localhost:5984";
const dbName = "mltest";

app.get('/', async (req, res) => {
    const authenticator = new BasicAuthenticator({
        username: username,
        password: password
    });
    const service = new CloudantV1({ authenticator });
    service.setServiceUrl(serviceUrl);    
    const docs = await service.postAllDocs({ db: dbName, includeDocs: true, limit: 50, skip: 0 });
    res.json(docs);
});

app.get('/_mem', async (req, res) => {
    res.json(memoryUsage());
});

app.get('/_gc', async (req, res) => {
    global.gc();
    res.send("ok");
});

app.listen(3333, () => {
    console.log('Application listening on port 3333!');
});
```

`load-test.js`
```js
import http from 'k6/http';

export default function () {
    http.get('http://localhost:3333');
}
```

Starting with a warm node process with _rss_ around 66M, _total heap_ at 14M and _used heap_ at 11M I hit it with load generator and pulled memory profile during the run. The _rss_ raised to 100M and _total heap_ to 48M. After the run I forced global gc on the node process and got memory values closed to initial.

Here is the graph for the whole test and it seems to be similar to yours with _rss_ growing rapidly and then staying stable under load.

<img width="832" alt="load-run" src="https://user-images.githubusercontent.com/70542/145262865-d516b57b-6471-4f24-92f7-32a05bd19bf8.png">

My test db is 1mil count with docs roughly 1k in size and in case you need a scale of load in test to compare to your case here is k6 summary:
```
> k6 run --vus 5 --duration 5m load-test.js
running (5m00.1s), 0/5 VUs, 18879 complete and 0 interrupted iterations
default ✓ [======================================] 5 VUs  5m0s

     data_received..................: 413 MB 1.4 MB/s
     data_sent......................: 1.5 MB 5.0 kB/s
     http_req_blocked...............: avg=9.64µs  min=1µs     med=3µs     max=23.7ms  p(90)=4µs     p(95)=5µs
     http_req_connecting............: avg=5.89µs  min=0s      med=0s      max=22.27ms p(90)=0s      p(95)=0s
     http_req_duration..............: avg=79.34ms min=14.67ms med=48.33ms max=1.97s   p(90)=72.66ms p(95)=141.22ms
       { expected_response:true }...: avg=79.34ms min=14.67ms med=48.33ms max=1.97s   p(90)=72.66ms p(95)=141.22ms
     http_req_failed................: 0.00%  ✓ 0         ✗ 18879
     http_req_receiving.............: avg=88.88µs min=27µs    med=87µs    max=3.2ms   p(90)=115µs   p(95)=128µs
     http_req_sending...............: avg=18.45µs min=6µs     med=18µs    max=1.03ms  p(90)=22µs    p(95)=25µs
     http_req_tls_handshaking.......: avg=0s      min=0s      med=0s      max=0s      p(90)=0s      p(95)=0s
     http_req_waiting...............: avg=79.24ms min=14.55ms med=48.22ms max=1.97s   p(90)=72.53ms p(95)=141.15ms
     http_reqs......................: 18879  62.917007/s
     iteration_duration.............: avg=79.44ms min=14.77ms med=48.42ms max=1.97s   p(90)=72.78ms p(95)=141.29ms
     iterations.....................: 18879  62.917007/s
     vus............................: 5      min=5       max=5
     vus_max........................: 5      min=5       max=5
``` 

To summarize I highly doubt there is a memory leak in sdk, it exhibits behavior typical to node app under load and respects garbage collector when invoked.

There might be some reference retention in your app as I'm sure it's more complex than this simple test. One thing I've noticed in your snippet is that you are initiating `CloudantV1` instance per request and that might be a better idea to share it instead, though I wouldn't expect this to make a drastic difference.

I'd recommend to try and use [node inspector](https://nodejs.org/en/docs/guides/debugging-getting-started/#enable-inspector) on your app to see where exactly memory goes.

**Comment by dtto86** [comment]
Hi @eiri 

We ran load test of 5K requests against two different ExpressJs apps, one based on `ibm-cloud-sdk-core` and the other on `@cloudant/cloudant`. We didn't observe any memory leak with/without calling global.gc() but observed higher latency in `ibm-cloud-sdk-core` based code when compared against `@cloudant/cloudant` based code.

// package.json
```
"dependencies": {
    "@ibm-cloud/cloudant": "0.0.22",
    "express": "^4.17.1",
    "ibm-cloud-sdk-core": "^2.17.0",
    "@cloudant/cloudant": "^4.3.0"
  }
```

// index.js based on ibm-cloud-sdk-core
```
const express = require('express');
const app = express();
const port = 3000;
const { BasicAuthenticator } = require('ibm-cloud-sdk-core');
const { CloudantV1 } = require('@ibm-cloud/cloudant');

const username = 'username';
const password = 'password';
const serviceUrl = 'service_url';

const authenticator = new BasicAuthenticator({
    username,
    password
});
const service = new CloudantV1({ authenticator });
service.setServiceUrl(serviceUrl);

app.get('/', async (req, res) => {
    try {
        const docs = await service.postAllDocs({ db: 'database_name', includeDocs: true, limit: 50, skip: 0 });
        res.json(docs);
    } catch (error) {
        console.log(error);
    }

});

app.get('/_mem', async (req, res) => {
    res.json(process.memoryUsage());
});

app.get('/_gc', async (req, res) => {
    global.gc();
    res.json('cleared..');
});

app.listen(port, () => {
    console.log(`App is running on port ${port}`);
});
```

// index-library.js based on @cloudant/cloudant
```
const express = require('express');
const app = express();
const port = 3000;
const cloudantDriver = require('@cloudant/cloudant');
const asyncMod = require('async');

const username = 'username';
const password = 'password';

const cloudant = new cloudantDriver({
    account: username,
    password: password
});
const db = cloudant.db.use('database_name');

app.get('/', async (req, resp) => {
    try {
        asyncMod.parallel([function (cb) {
            // all design docs will appear in a group. endkey:_ will return docs upto the first ddoc
            const reqParams = {
                endkey: '_',
                include_docs: true
            }
            db.list(reqParams, function (err, res, resHeaders) {
                if (err) cb(err)
                else cb(null, res.rows, { reqUri: resHeaders.uri, reqParams: reqParams })
            })
        }, function (cb) {
            // startkey:` will return documents after the last design document
            const reqParams = {
                startkey: '`',
                include_docs: true
            }
            db.list(reqParams, function (err, res, resHeaders) {
                if (err) cb(err)
                else cb(null, res.rows, { reqUri: resHeaders.uri, reqParams: reqParams })
            })
        }
        ], function (err, response, headers) {
            if (err) {
                console.log(err);
            }
            resp.json(response);
        });
    } catch (error) {
        console.log(error);
    }

});

app.get('/_mem', async (req, res) => {
    res.json(process.memoryUsage());
});

app.get('/_gc', async (req, res) => {
    global.gc();
    res.json('cleared..');
});

app.listen(port, () => {
    console.log(`App is running on port ${port}`);
});
```

// loadtest command
```
loadtest -n 5000 -c 5 http://localhost:3000/
```

// ibm-cloud-sdk-core: loadtest command response
```
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO Target URL:          http://localhost:3000
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO Max requests:        5000
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO Concurrency level:   5
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO Agent:               none
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO Completed requests:  5000
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO Total errors:        0
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO Total time:          836.3058089 s
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO Requests per second: 6
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO Mean latency:        835.6 ms
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO Percentage of the requests served within a certain time
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO   50%      754 ms
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO   90%      1125 ms
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO   95%      1256 ms
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO   99%      1990 ms
[Mon Dec 13 2021 15:18:51 GMT+0530 (India Standard Time)] INFO  100%      12490 ms (longest request)
```

// @cloudant/cloudant: loadtest command response
```
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO Target URL:          http://localhost:3000
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO Max requests:        5100
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO Concurrency level:   1
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO Agent:               none
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO Completed requests:  5100
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO Total errors:        0
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO Total time:          977.9648994 s
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO Requests per second: 5
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO Mean latency:        191.7 ms
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO Percentage of the requests served within a certain time
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO   50%      179 ms
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO   90%      197 ms
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO   95%      217 ms
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO   99%      511 ms
[Mon Dec 13 2021 16:27:06 GMT+0530 (India Standard Time)] INFO  100%      2723 ms (longest request)
```

Is there a way to optimize the latency in response?

**Comment by eiri** [comment]
Hi @dtto86 

I see you've used different concurrency levels on old and new sdks tests, so you data is not really comparable. On top of that old library was using `GET` for that type of request and new one using `POST` with payload compression, so they are not going to match one to one in any case.

In general one-off latency tests are highly volatile and not representative of anything, a proper way to do such testing would be a multiple runs over isolated back-end with identical starting state to accumulate a statistically significant amount of data.

In general there is no generic way to influence the latency times, it's very dependent on back-end, network and application architecture and need to be solved on case-by-case basis.

Since we agreed that there is no memory leak this issue was initially opened about, I'm going to close it now.
